ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,676 @@
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE DeriveGeneric        #-}
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE NamedFieldPuns       #-}
+{-# LANGUAGE RankNTypes           #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE StandaloneDeriving   #-}
+{-# LANGUAGE TupleSections        #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | ChainSync jumping (CSJ) is an optimization for the ChainSync protocol that
+-- allows nodes to sync without downloading headers from all of the honest
+-- peers. This load is undesirable as it slows down all the peers involved.
+--
+-- The idea is to download the headers of a chain from a single peer (the
+-- dynamo) and then ask periodically to the other peers (the jumpers) whether
+-- they agree with the dynamo's chain.
+--
+-- When the jumpers disagree with the dynamo, the jumper with the oldest
+-- intersection is asked to compete with the dynamo in the GDD logic. If the
+-- dynamo is disconnected, a new dynamo is elected and the objector is demoted
+-- to a jumper.
+--
+-- If the objector is disconnected, the syncing process continues with the
+-- dynamo and the remaining jumpers.
+--
+-- The main property of the algorithm to avoid network load, is that it never
+-- dowloads headers from more than two plausibly honest peers at a time (a
+-- dynamo and an objector). All other peers are either waiting their turn to
+-- compete with the dynamo, or are in agreement with it, or are disengaged.
+--
+-- Interactions with the GDD
+-- -------------------------
+--
+-- It is possible that neither the dynamo nor the objector are disconnected.
+-- This could happen if:
+-- 1. They both server the same chain
+-- 2. They both claim to have no more headers
+--
+-- To avoid (1) CSJ checks that the objector disagrees with the dynamo at the
+-- point it claimed to disagree as a jumper. If the objector agrees with the
+-- dynamo, it is disengaged. A disengaged peer is not asked to jump or act as
+-- dynamo or objector. Instead, it continues to offer headers for the rest of
+-- the syncing. When the objector is disengaged, a new objector is elected
+-- among the dissenting jumpers. If there are no dissenting jumpers left, the
+-- syncing continues with the dynamo and the remaining jumpers.
```

Comment:
> Or is the idea that this situation could only happen on the historical chain with adversarial peers because it cannot happen without rollbacks?

I don't think peers can be disengaged unless they do historical rollbacks, or they do honest rollbacks when the node is almost caught up.